From fab51608a0acdb2e91620e102998d4d89c63ae02 Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Wed, 8 Jun 2011 17:18:59 +0200
Subject: [PATCH 042/285] ChannelScan: Update multiplex transponder and
 network id when channel updated

When a transport is added manually using the transport editor, the transport
and network ids are not known and are left unset.  This causes a problem
for interactive TV which tunes a channel by network and transport id.

This patch updates the transport and network ids for a multiplex whenever a
channel is found or updated on the transport.

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
---
 .../libs/libmythtv/channelscan/channelimporter.cpp |   31 ++++++++++++++++++--
 1 file changed, 28 insertions(+), 3 deletions(-)

diff --git a/mythtv/libs/libmythtv/channelscan/channelimporter.cpp b/mythtv/libs/libmythtv/channelscan/channelimporter.cpp
index 1bce802..2621b7d 100644
--- a/mythtv/libs/libmythtv/channelscan/channelimporter.cpp
+++ b/mythtv/libs/libmythtv/channelscan/channelimporter.cpp
@@ -477,15 +477,28 @@ ScanDTVTransportList ChannelImporter::InsertChannels(
 
                 chan.channel_id = (chanid > 0) ? chanid : chan.channel_id;
 
-                if (chan.channel_id && !chan.db_mplexid)
+                if (chan.channel_id)
                 {
                     uint tsid = chan.vct_tsid;
                     tsid = (tsid) ? tsid : chan.sdt_tsid;
                     tsid = (tsid) ? tsid : chan.pat_tsid;
                     tsid = (tsid) ? tsid : chan.vct_chan_tsid;
 
-                    chan.db_mplexid = ChannelUtil::CreateMultiplex(
-                        chan.source_id, transports[i], tsid, chan.orig_netid);
+                    if (!chan.db_mplexid)
+                    {
+                        chan.db_mplexid = ChannelUtil::CreateMultiplex(
+                            chan.source_id, transports[i], tsid, chan.orig_netid);
+                    }
+                    else
+                    {
+                        // Find the matching multiplex. This updates the
+                        // transport and network ID's in case the transport
+                        // was created manually
+                        int id = ChannelUtil::GetBetterMplexID(chan.db_mplexid,
+                                    tsid, chan.orig_netid);
+                        if (id >= 0)
+                            chan.db_mplexid = id;
+                    }
                 }
 
                 if (chan.channel_id && chan.db_mplexid)
@@ -634,6 +647,18 @@ ScanDTVTransportList ChannelImporter::UpdateChannels(
 
                 ChannelUtil::UpdateInsertInfoFromDB(chan);
 
+                // Find the matching multiplex. This updates the
+                // transport and network ID's in case the transport
+                // was created manually
+                uint tsid = chan.vct_tsid;
+                tsid = (tsid) ? tsid : chan.sdt_tsid;
+                tsid = (tsid) ? tsid : chan.pat_tsid;
+                tsid = (tsid) ? tsid : chan.vct_chan_tsid;
+                int id = ChannelUtil::GetBetterMplexID(chan.db_mplexid,
+                            tsid, chan.orig_netid);
+                if (id >= 0)
+                    chan.db_mplexid = id;
+
                 updated = ChannelUtil::UpdateChannel(
                     chan.db_mplexid,
                     chan.source_id,
-- 
1.7.9.5

